File: public/download/years/2011/sada1/opt.py
import sys

# Computation of a really sophisticated function.
# Programmed by the Interlos organizing team.
# All rights reserved.

def f(A, B, C):
    s = 0
    for i in range(0, C):
        s = (s + A * g(B, i, h(i))) % h(i+1)
    return s % h(C)

def g(a, b, c):
    r = 1
    for i in range(0, b):
        r = (r * a) % c
    return r

def h(i):
    p = 10000*[None]
    pc = 1
    p[1] = 2
    while True:
        n = p[pc]
        ok = False
        while True:
            thisOk = True
            n += 1
            for j in range(1, pc+1):
                if n % p[j] == 0: thisOk = False
            if thisOk: ok = True

            if ok: break

        pc += 1
        p[pc] = n

        if pc >= 9999: break

    return p[ m(i) ]

def m(i):
    k = 6543 ^ i
    for n in range(0, (3456 & i)):
        for j in range(0, i):
            k = k ^ j
    return i ^ k


inValues = sys.stdin.readline().split(' ')
A = int(inValues[0].strip())
B = int(inValues[1].strip())
C = int(inValues[2].strip())
print(f(A, B, C))

File: public/download/years/2011/sada1/opt.c
/**
 * Computation of a really sophisticated function.
 * Programmed by the Interlos organizing team.
 * All rights reserved.
 */
#include <stdio.h>

int f(int A, int B, int C);
int g(int a, int b, int c);
int h(int i);
int m(int i);

int main()
{
    int A, B, C;
    scanf("%d %d %d", &A, &B, &C);
    printf("%d\n", f(A, B, C));
    return 0;
}


int f(int A, int B, int C)
{
    int i;
    int s = 0;
    for (i = 0; i < C; i++) {
        s = (s + A * g(B, i, h(i))) % h(i+1);
    }
    return s % h(C);
}


int g(int a, int b, int c)
{
    int i;
    int r = 1;
    for (i = 0; i < b; i++) {
        r = (r * a) % c;
    }
    return r;
}


int h(int i)
{
    int p[10000];
    int pc = 1;
    p[1] = 2;
    do {
        int n = p[pc];
        int ok = 0;
        do {
            int j;
            int thisOk = 1;
            n++;
            for (j = 1; j <= pc; j++) {
                if (n % p[j] == 0) thisOk = 0;
            }
            if (thisOk) ok = 1;
        } while (!ok);

        pc++;
        p[pc] = n;
    } while (pc < 9999);
    return p[ m(i) ];
}


int m(int i)
{
    int k = 6543 ^ i;
    int j, n;
    for (n = 0; n < (3456 & i); n++) {
        for (j = 0; j < i; j++) {
            k = k ^ j;
        }
    }
    return i ^ k;
}
